### PR TITLE
refactor(layered-drgporep): restructure parallelized proving

### PR DESCRIPTION
**Pull request by dignifiedquire**
Perf is usually a bit better than master for me, but variance is too high for me to be sure. 

- no arcs or cell types
- only `rayon::join` (meaning it uses a work stealing thread pool under the hood)
- no channels
- no reliance on `panic`, but returning `Result` types
- a little less cloning and allocations than before
- 2/3 of the code

### CODE REVIEW

File: storage-proofs/src/layered_drgporep.rs [hunk 1]
```diff
@@ -356,96 +354,63 @@ pub trait Layers {
...
+                                        &current_drgpp.graph,
+                                        current_drgpp.sloth_iter,
+                                        replica_id,
+                                        data,
+                                    )
+                                } else {
+                                    Ok(())
+                                }
+                            },
+                            || Self::transform(&current_drgpp, layer, layers),
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/layered_drgporep.rs]: I did get speedups when doing this vs leaving this join out interestingly enough

File: storage-proofs/src/layered_drgporep.rs [hunk 2]
```diff
@@ -356,96 +354,63 @@ pub trait Layers {
...
+                        );
+
+                        encoded.map(|_| params)
+                    },
+                );
+
+                trees.push(tree_d?);
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs/src/layered_drgporep.rs]: You are reading correct. I do not think given that merkletrees are generated in parallel this actually helps us much, and my tests indicated as much, but I am not certain. I pushed a version that now uses work stealing queue instead, based on crossbeam.

I would ask you to use your experience in running larger things to run these two versions on configs you used before and let me know how things look.

**Comment by porcuquine** [comment]
Some quick notes:
 - Should rename this PR, since current name doesn't apply anymore.
 - Unfortunately, when I tested this I saw a relatively small but repeatably noticeable decrease in performance.

I'm fine with keeping this around as something which might bear fruit, but if it's going to require per-system tuning, let's use the config mechanism and ensure there's a way to avoid regressions for the default settings. (I'm a little concerned that this will be hard unless the whole feature is opt-in.)

Basically, while I agree that this feels cleaner than what it replaces, we need to see that it's not just spreading complexity and/or has a measurable perf benefit.

**Comment by dignifiedquire** [comment]
closing for now
